File: 729.my-calendar-i.java
/*
 * @lc app=leetcode id=729 lang=java
 *
 * [729] My Calendar I
 */
import java.util.*;

// @lc code=start
class MyCalendar {
    private TreeSet<int[]> bookings;

    public MyCalendar() {
        bookings = new TreeSet<int[]>((int[] a, int[] b) -> a[0] - b[0]);
    }

    // O(log(n))
    public boolean book(int start, int end) {
        int[] booking = new int[] {start, end};
        int[] ceil = bookings.ceiling(booking);
        if (ceil != null && ceil[0] < end) return false;
        int[] floor = bookings.floor(booking);
        if (floor != null && start < floor[1]) return false; 
        bookings.add(booking);
        return true;
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */
// @lc code=end

File: 729.my-calendar-i.cpp
/*
 * @lc app=leetcode id=729 lang=cpp
 *
 * [729] My Calendar I
 */
#include <bits/stdc++.h>

// @lc code=start
typedef std::pair<int, int> booking;

// S: O(n)
class MyCalendar {
    std::set<booking> bookings;

public:
    MyCalendar() {}

    // T: O(log(n))
    bool book(int start, int end) {
        booking b(start, end);
        auto next = bookings.upper_bound(b);
        if (next != bookings.end() && next->first < end) {
            // equivalent to Java's TreeMap.ceiling()
            // there is an existing interval with start < the booking's end
            return false; 
        }
        if (next != bookings.begin() && start < std::prev(next)->second) {
            // equivalent to Java's TreeMap.floor()
            // there is an existing interval with end > the booking's start
            return false; 
        }
        bookings.insert(std::move(b));
        return true;
    }
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */
// @lc code=end

